core/src/main/java/de/jplag/merging/MatchMerging.java

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * Constructor for class MatchMerging
+     * @param result is the initially computed result object
+     * @param options encapsulates the adjustable options
+     */
+    public MatchMerging(JPlagResult result, JPlagOptions options) {
+        this.result = result;
+        this.comparisons = new ArrayList<>(result.getAllComparisons());
+        this.options = options;
+        this.minimumTokenMatch = options.minimumTokenMatch();
+        this.seperatingThreshold = options.mergingParameters().seperatingThreshold();
+    }
+
+    /**
+     * Runs the internal match merging pipeline. It computes neighboring matches, merges them based on
+     * {@link MergingParameters} and removes the merge buffer afterwards.
+     * @return JPlagResult containing the merged matches
+     */
+    public JPlagResult run() {
+        long timeBeforeStartInMillis = System.currentTimeMillis();
+        for (int i = 0; i < comparisons.size(); i++) {
+            firstSubmission = comparisons.get(i).firstSubmission().copy();
+            secondSubmission = comparisons.get(i).secondSubmission().copy();
+            globalMatches = new ArrayList<>(comparisons.get(i).matches());
+            globalMatches.addAll(comparisons.get(i).ignoredMatches());
+            computeNeighbors();
+            mergeNeighbors();
+            removeBuffer();
+            comparisons.set(i, new JPlagComparison(firstSubmission, secondSubmission, globalMatches, new ArrayList<>()));
+
+        }
+        long durationInMillis = System.currentTimeMillis() - timeBeforeStartInMillis;
+        return new JPlagResult(comparisons, result.getSubmissions(), result.getDuration() + durationInMillis, options);
+    }
+
+    private void computeNeighbors() {
+        neighbors = new ArrayList<>();
+        List<Match> sortedByFirst = new ArrayList<>(globalMatches);
+        Collections.sort(sortedByFirst, (m1, m2) -> m1.startOfFirst() - m2.startOfFirst());
+        List<Match> sortedBySecond = new ArrayList<>(globalMatches);
+        Collections.sort(sortedBySecond, (m1, m2) -> m1.startOfSecond() - m2.startOfSecond());
+        for (int i = 0; i < sortedByFirst.size() - 1; i++) {
+            if (sortedBySecond.indexOf(sortedByFirst.get(i)) == (sortedBySecond.indexOf(sortedByFirst.get(i + 1)) - 1)) {
+                neighbors.add(Arrays.asList(sortedByFirst.get(i), sortedByFirst.get(i + 1)));
+            }
+        }
+    }
+
+    private void mergeNeighbors() {
+        int i = 0;
+        while (i < neighbors.size()) {
+            int lengthUpper = neighbors.get(i).get(0).length();
+            int lengthLower = neighbors.get(i).get(1).length();
+            int seperatingFirst = neighbors.get(i).get(1).startOfFirst() - neighbors.get(i).get(0).endOfFirst() - 1;
+            int seperatingSecond = neighbors.get(i).get(1).startOfSecond() - neighbors.get(i).get(0).endOfSecond() - 1;
+            double seperating = (seperatingFirst + seperatingSecond) / 2.0;
```

Comment:
What does separating mean here?

Reply:
Separation? TokensBetweenNeighbors?

Reply:
seperating is the average of how many tokens separate neighbouring matches in the first submission and how many separate them in the second

Reply:
Would renaming it TokenBetweenNeighbours resolve this for you?

Reply:
How about naming `separating` to `averageTokensBetweenMatches`? I think that helps to understand it better.
the `seperatingFirst` and `seperatingSecond` could be `tokenBetweenFirst` and `tokensBetweenSecond`.
I think one problem is that first/second is ambiguous, because it can both refer to the matches of a single token sequence, but also to the two sequences in the comparison itself. One idea would be to use left/right for the token sequences, and the first, second etc. only regarding the matches in a sequence and their relation.

Reply:
The ambiguity was the reason I initially named it left, right, upper and lower.
But I changed that due to a request from Alex.